From 2b23fa9f000c410875b09fe8b6a5c77a94384ba2 Mon Sep 17 00:00:00 2001
From: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
Date: Mon, 22 Dec 2025 02:55:41 +0100
Subject: [PATCH 96/98] net: phy: smsc: Add support for 25MHz clock input and
 RMII REFCLKO out - WiP

TODO: commit message

Signed-off-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>

[dts: rebased onto 6.6.144 for 6.6-lts - devm_clk_get_optional_enabled_with_rate()
 doesn't exist at this baseline; kept the split
 devm_clk_get_optional_enabled() + clk_set_rate() this file already used,
 with the computed in_clk_mhz]
---
diff --git a/drivers/net/phy/smsc.c b/drivers/net/phy/smsc.c
index 5f9bb0ebe..7eb1496ae 100644
--- a/drivers/net/phy/smsc.c
+++ b/drivers/net/phy/smsc.c
@@ -13,6 +13,7 @@
  */
 
 #include <linux/clk.h>
+#include <linux/clk-provider.h>
 #include <linux/kernel.h>
 #include <linux/module.h>
 #include <linux/mii.h>
@@ -657,11 +658,35 @@ int smsc_phy_set_tunable(struct phy_device *phydev,
 }
 EXPORT_SYMBOL_GPL(smsc_phy_set_tunable);
 
+static int smsc_phy_register_rmii_refclk_output(struct device *dev)
+{
+	struct clk_hw *refclko;
+	char *name;
+
+	if (!IS_ENABLED(CONFIG_COMMON_CLK))
+		return 0;
+
+	name = devm_kasprintf(dev, GFP_KERNEL, "%s-refclko", dev_name(dev));
+	if (!name)
+		return -ENOMEM;
+
+	refclko = devm_clk_hw_register_fixed_factor_index(dev, name, 0, 0,
+							  2, 1);
+	if (IS_ERR(refclko))
+		return dev_err_probe(dev, PTR_ERR(refclko),
+				     "Failed to register RMII REF_CLK clock output\n");
+
+	return devm_of_clk_add_hw_provider(dev, of_clk_hw_simple_get, refclko);
+}
+
 int smsc_phy_probe(struct phy_device *phydev)
 {
 	struct device *dev = &phydev->mdio.dev;
 	struct smsc_phy_priv *priv;
+	bool has_rmii_ref_clk_in;
+	unsigned long in_clk_mhz;
 	struct clk *refclk;
+	int rc;
 
 	priv = devm_kzalloc(dev, sizeof(*priv), GFP_KERNEL);
 	if (!priv)
@@ -675,13 +700,33 @@ int smsc_phy_probe(struct phy_device *phydev)
 
 	phydev->priv = priv;
 
+	has_rmii_ref_clk_in = !device_property_present(dev, "clock-names") ||
+			      device_property_match_string(dev, "clock-names",
+							   "ref_clk") >= 0;
+	if (has_rmii_ref_clk_in)
+		in_clk_mhz = 50;
+	else
+		in_clk_mhz = 25;
+
 	/* Make clk optional to keep DTB backward compatibility. */
 	refclk = devm_clk_get_optional_enabled(dev, NULL);
 	if (IS_ERR(refclk))
 		return dev_err_probe(dev, PTR_ERR(refclk),
-				     "Failed to request clock\n");
+				     "Failed to request %luMHz clock input\n",
+				     in_clk_mhz);
 
-	return clk_set_rate(refclk, 50 * 1000 * 1000);
+	rc = clk_set_rate(refclk, in_clk_mhz * 1000 * 1000);
+	if (rc)
+		return rc;
+
+	if (phydev->interface == PHY_INTERFACE_MODE_RMII &&
+	    !has_rmii_ref_clk_in) {
+		rc = smsc_phy_register_rmii_refclk_output(dev);
+		if (rc)
+			return rc;
+	}
+
+	return 0;
 }
 EXPORT_SYMBOL_GPL(smsc_phy_probe);
 
